codec.go

```diff
@@ -56,20 +59,43 @@ func setupCodec() error {
 				max = addr
 			}
 		}
+		addrs = append(addrs, int(addr))
 	}
 	addrRange := max - min
 	if addrRange == 0 {
 		return fmt.Errorf("failed to get address range of types")
 	}
-	if addrRange > maxAcceptableTypeAddrRange {
+	sort.Ints(addrs)
+	typeAddrShift = calcTypeOffsetShift(addrs)
+	cacheSize := addrRange >> typeAddrShift
+	if cacheSize > maxAcceptableTypeAddrRange {
 		return fmt.Errorf("too big address range %d", addrRange)
 	}
-	cachedDecoder = make([]decoder, addrRange)
+	cachedDecoder = make([]decoder, cacheSize)
 	baseTypeAddr = min
 	maxTypeAddr = max
 	return nil
 }
 
+func calcTypeOffsetShift(addrs []int) uintptr {
+	offset64 := true
+	offset32 := true
+	for i := 1; i < len(addrs); i++ {
+		if incr := addrs[i] - addrs[i-1]; incr&63 != 0 {
+			offset64 = false
+		} else if incr&31 != 0 {
```

Comment:
This is incorrect, `incr&63 != 0` does not exclude `incr&31 != 0`, I will fix this.